nncf/experimental/torch/nas/bootstrapNAS/search/search.py

```diff
@@ -0,0 +1,492 @@
+
+        return cls(num_evals, num_constraints, population,
+                   seed, crossover_prob, crossover_eta,
+                   mutation_prob, mutation_eta, acc_delta, ref_acc)
+
+
+class BaseSearchAlgorithm:
+    """
+    Base class for search algorithms. It contains the evaluators used by search approches.
+    """
+    def __init__(self):
+        """
+        Initialize BaseSearchAlgorithm class.
+
+        """
+        self._use_default_evaluators = True
+        self._evaluator_handlers = []
+        self._accuracy_evaluator_handler = None
+        self._efficiency_evaluator_handler = None
+        self._log_dir = None
+        self._search_records = []
+        self.bad_requests = []
+        self.best_config = None
+        self.best_vals = None
+        self.best_pair_objective = float('inf')
+        self._tb = None
+
+    @property
+    def search_records(self):
+        return self._search_records
+
+    @abstractmethod
+    def run(self, validate_fn: Callable, val_loader: DataLoader, checkpoint_save_dir: str,
+            efficiency_evaluator: Optional[BaseEvaluator] = None, ref_acc: Optional[float] = 100,
+            tensorboard_writer: Optional[SummaryWriter] = None) -> Tuple[ElasticityController,
+                                                                         SubnetConfig, Tuple[float, ...]]:
+        """This method should implement how to run the search algorithm."""
+
+    def search_progression_to_csv(self, filename='search_progression.csv') -> NoReturn:
+        """
+        Exports search progression to CSV file
+
+        :param filename: path to save the CSV file.
+        :return:
+        """
+        with open(f'{self._log_dir}/{filename}', 'w', encoding='utf8') as progression:
+            writer = csv.writer(progression)
+            for record in self.search_records:
+                writer.writerow(record)
+
+
+class SearchAlgorithm(BaseSearchAlgorithm):
+    def __init__(self,
+                 model: NNCFNetwork,
+                 elasticity_ctrl: ElasticityController,
+                 nncf_config: NNCFConfig,
+                 verbose=True):
+        """
+        Initializes search algorithm
+
+        :param model: Super-network
+        :param elasticity_ctrl: interface to manage the elasticity of the super-network.
+        :param nncf_config: Configuration file.
+        :param verbose:
+        """
+        super().__init__()
+        self._model = model
+        self._elasticity_ctrl = elasticity_ctrl
+        self._elasticity_ctrl.multi_elasticity_handler.activate_maximum_subnet()
+        search_config = nncf_config.get('bootstrapNAS', {}).get('search', {})
+        self.num_obj = None
+        self.search_params = SearchParams.from_dict(search_config)
+        self._log_dir = nncf_config.get("log_dir", ".")
+        self._verbose = verbose
+        self._top1_accuracy_validation_fn = None
+        self._val_loader = None
+        evo_algo = search_config['algorithm']
```

Comment:
it fails with no trivial error if there's no search algorithm.
```
evo_algo = search_config['algorithm']
E           KeyError: 'algorithm'
```
I'd use a default value:
evo_algo = search_config.get('algorithm', EvolutionaryAlgorithms.NSGA2.value)

Reply:
Done